### DISCUSSION TITLE
the issue of Unity graphics acceleration

### DISCUSSION THREAD

**Discussion by TomatoPicker278** [empriselab/RCareWorld#165] [Ideas / Suggestions]
I would like to use our lab's Nvidia GPU server to accelerate the animation of the bathing scenario on my personal computer, but I haven’t used Unity before, so I would like to ask about some technical issues regarding graphics acceleration.

Do I need to install the Unity software on the server and configure the necessary graphics acceleration settings, then connect my local computer to the server through SSH? Do I also need to add relevant code in `test_bathing.py` or configure Unity settings on my local machine?

Is there any documentation on Unity graphics acceleration? Could you please share it?

Thank you for your help 🙏

**Comment by YoruCathy** [comment]
Unfortunately, our simulation mostly runs on CPU. We do notice if you have Nvidia driver installed properly it might run in a more stable way. So I'd suggest follow the pre-install guide and install these libraries and the Nvidia driver.

**Reply by TomatoPicker278** [reply]
No problem, thank you for your response. I just wanted to confirm once again, when I run the `test_bathing.py` program, is the underlying engine it calls Unity? If so, I can try to connect my personal computer with the server for testing.

**Reply by YoruCathy** [reply]
It calls Unity, which is the executable file included in the codebase. Can you explain more about what you mean by connecting your personal computer with the server?

**Reply by TomatoPicker278** [reply]
Our lab has a server equipped with an Intel i7 processor and an Nvidia 4070 graphics card, with the Nvidia drivers already installed. When I run the `test_bathing.py` program on the lab’s server, it automatically uses Nvidia’s graphics acceleration, resulting in very smooth visuals.

However, due to certain reasons, I currently need to run the `test_bathing.py` program on my office computer, which does not have a dedicated graphics card. This causes the animation to run very slowly and even lag, so I need to connect my computer to the server.

I attempted to directly connect my personal computer to the server via SSH and ran the `test_bathing.py` program on my local computer. However, there was still no graphics acceleration, and the visuals were still laggy, indicating that the server’s GPU acceleration was not being utilized.

I am currently considering whether the issue is because I have not connected the Unity engine on my local computer (where `test_bathing.py` is running) with the Unity on the server. I’m thinking that if I could connect the two Unity instances, it might resolve my issue.

That’s why I asked you my previous question.

**Reply by YoruCathy** [reply]
First, make sure it is running the Unity instance on the server. It sounds like it is still calling the instance on your local computer. Actually, is it possible to just run the entire code on the server? You can always code on your local machine and run it on the server. Or is there anything specific that requires two computers to run it?

**Reply by TomatoPicker278** [reply]
I've looked into this issue recently. The core of this problem is figuring out how to offload image rendering tasks from the local computer to a remote GPU. So far, I've found one solution that uses a combination of VirtualGL and TurboVNC. VirtualGL transfers the graphics acceleration tasks to the server’s GPU, while TurboVNC streams the remote desktop to the local computer. However, to be honest, this method is quite complex, so I currently still recommend running the program directly on the server, as it will be much more convenient.

**Reply by YoruCathy** [reply]
I see. I guess other than streaming the rendering using the methods you mentioned, there are no other good solutions that I'm aware of. The server running eval ai on our end has graphics rendering support, so when you submit it, it should be able to run on a single desktop.
